activerecord/lib/active_record/test_fixtures.rb

```diff
@@ -173,10 +173,33 @@ def teardown_fixtures
     end
 
     def enlist_fixture_connections
+      setup_shared_connection_pool
+
       ActiveRecord::Base.connection_handler.connection_pool_list.map(&:connection)
     end
 
     private
+
+      # Shares the writing connection pool with connections on
+      # other handlers.
+      #
+      # In an application with a primary and replica the test fixtures
+      # need to share a connection pool so that the reading connection
+      # can see data in the open transaction on the writing connection.
+      def setup_shared_connection_pool
+        writing_handler = ActiveRecord::Base.connection_handler
+
+        ActiveRecord::Base.connection_handlers.values.each do |handler|
+          if handler != writing_handler
+            handler.connection_pool_list.each do |pool|
+              name = pool.spec.name
+              writing_connection = writing_handler.retrieve_connection_pool(name)
+              handler.send(:owner_to_pool)[name] = writing_connection
```

Comment:
Why do we need to share the writing connection pool for all handlers?

I've found the change when creating the example at https://github.com/rails/rails/pull/35073#issuecomment-458519770, since the `TestFixtures` is included in the `ActiveSupport::TestCase`, the overwriting connection pool by the writing connection pool is happened in the `before_setup`, so it is hard to testing replica connections in their apps.

https://github.com/rails/rails/blob/bf42e1506be85c0c977038ae334548f5699f7afd/railties/lib/rails/test_help.rb#L23-L29

Unless suppressing the effect of `setup_shared_connection_pool`, the following test won't be passed.

```diff
diff --git a/test/test_helper.rb b/test/test_helper.rb
index 0ff12e7..5d82a8b 100644
--- a/test/test_helper.rb
+++ b/test/test_helper.rb
@@ -10,4 +10,11 @@ class ActiveSupport::TestCase
   fixtures :all
 
   # Add more helper methods to be used by all tests here...
+
+  # `enlist_fixture_connections` replaces connection pools in non-default handlers
+  # by default writer connection pool.
+  # We can't test `:reading` connection unless suppressing the effect of the method for now.
+  def enlist_fixture_connections
+    []
+  end
 end
```

```diff
diff --git a/test/models/post_test.rb b/test/models/post_test.rb
index 6d9d463..f2523ce 100644
--- a/test/models/post_test.rb
+++ b/test/models/post_test.rb
@@ -1,6 +1,11 @@
 require 'test_helper'
 
 class PostTest < ActiveSupport::TestCase
+  test "replica?" do
+    pool = Post.connection_handlers[:reading].connection_pool_list.first
+    assert pool.connection.replica?
+  end
+
   # test "the truth" do
   #   assert true
   # end
```

Reply:
If we don't combine them, then one connection will be inside the fixture transaction and see the fixtures (& any changes to them), and the other will see empty tables.

Reply:
I see... so for now probably people need to restore (re-overwrite) pools in handlers if their app uses read-write-splitting or sharding and want to test that explicitly.

Reply:
Yeah, that approach sounds good to me.